explorer/interpreter/type_checker.cpp

```diff
@@ -2920,6 +2920,354 @@ auto TypeChecker::TypeCheckPattern(
   }
 }
 
+// A abstracted view of a pattern or constant value (which we view as a
+// particular kind of pattern).
+class AbstractPattern {
+ public:
+  enum Kind {
+    // A pattern that matches anything.
+    Wildcard,
+    // A pattern that matches a compound value with sub-patterns to match
+    // elements. A compound value is modeled as a discriminator name applied to
+    // a sequence of nested values: the alternative `Optional.Element(E)` has
+    // discriminator `Element` and nested value `E`, and the tuple `(A,B,C)`
+    // has an empty discriminator and nested values `A`, `B`, and `C`.
+    Compound,
+    // A pattern that matches a particular primitive value.
+    Primitive
+  };
+
+  AbstractPattern(Nonnull<const Pattern*> pattern) { Set(pattern); }
+
+  AbstractPattern(Nonnull<const Value*> value, Nonnull<const Value*> type)
+      : value_(value), type_(type) {}
+
+  // Make a match-anything wildcard pattern.
+  static auto MakeWildcard() -> AbstractPattern {
+    return AbstractPattern(WildcardTag());
+  }
+
+  auto kind() const -> Kind {
+    if (auto* pattern = value_.dyn_cast<const Pattern*>()) {
+      return Compound;
+    }
+    if (auto* value = value_.dyn_cast<const Value*>()) {
+      if (isa<TupleValue, AlternativeValue>(value)) {
+        return Compound;
+      }
+      return Primitive;
+    }
+    assert(value_.is<const WildcardTag*>());
```

Comment:
Should your `assert` uses generally be `CARBON_CHECK`?

Reply:
Done.